universal-application-tool-0.0.1/app/controllers/applicant/ApplicantInformationController.java

```diff
@@ -0,0 +1,115 @@
+package controllers.applicant;
+
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+
+import auth.ProfileUtils;
+import controllers.CiviFormController;
+import forms.ApplicantInformationForm;
+import java.util.Locale;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import javax.inject.Inject;
+import models.Applicant;
+import play.data.Form;
+import play.data.FormFactory;
+import play.i18n.MessagesApi;
+import play.libs.concurrent.HttpExecutionContext;
+import play.mvc.Http;
+import play.mvc.Result;
+import play.mvc.Results;
+import repository.ApplicantRepository;
+import services.applicant.ApplicantData;
+import services.applicant.ApplicantNotFoundException;
+import views.applicant.ApplicantInformationView;
+
+/**
+ * Provides methods for editing and updating an applicant's information, such as their preferred
+ * language.
+ */
+public final class ApplicantInformationController extends CiviFormController {
+
+  private final HttpExecutionContext httpExecutionContext;
+  private final MessagesApi messagesApi;
+  private final ApplicantInformationView informationView;
+  private final ApplicantRepository repository;
+  private final FormFactory formFactory;
+  private final ProfileUtils profileUtils;
+
+  @Inject
+  public ApplicantInformationController(
+      HttpExecutionContext httpExecutionContext,
+      MessagesApi messagesApi,
+      ApplicantInformationView informationView,
+      ApplicantRepository repository,
+      FormFactory formFactory,
+      ProfileUtils profileUtils) {
+    this.httpExecutionContext = httpExecutionContext;
+    this.messagesApi = messagesApi;
+    this.informationView = informationView;
+    this.repository = repository;
+    this.formFactory = formFactory;
+    this.profileUtils = profileUtils;
+  }
+
+  public CompletionStage<Result> edit(Http.Request request, long applicantId) {
+    return checkApplicantAuthorization(profileUtils, request, applicantId)
+        .thenApplyAsync(
+            v -> ok(informationView.render(request, applicantId)), httpExecutionContext.current())
+        .exceptionally(
+            ex -> {
+              if (ex instanceof CompletionException) {
+                if (ex.getCause() instanceof SecurityException) {
+                  return unauthorized();
+                }
+              }
+              throw new RuntimeException(ex);
+            });
+  }
+
+  public CompletionStage<Result> update(Http.Request request, long applicantId) {
+    Form<ApplicantInformationForm> form = formFactory.form(ApplicantInformationForm.class);
+    if (form.hasErrors()) {
```

Comment:
I'm not sure - @Bion suggested giving a `badRequest` for form errors, but I'm not sure I'm doing it correctly. Bion, any thoughts? `form.bindFromRequest` doesn't throw anything obvious, so I'm not sure how to catch issues here

Reply:
I used this from the Play docs: https://www.playframework.com/documentation/2.8.x/JavaForms#Handling-binding-failure

Reply:
This looks right to me!

The case to handle here is if someone POSTS form contents to this endpoint that is incorrectly formed we want to make sure to identify that as a 4XX-level error and not an internal server error. Calling `get()` will throw an exception if it can't bind the values it expects, so this guards against that.

Reply:
Thanks Bion!

Reply:
Thanks for the explanations. @carolinedanzi, I'm with you in that I'd expect the error to be thrown/caught after the call to `bindFromRequest`... But given that it doesn't throw something we can obviously catch, this lgtm.